### ISSUE TITLE
Improvement request: Branch instead of Fork

### ISSUE THREAD

**Issue by drew2a** [Tribler/tribler#5575]
## TLDR

Historically, Tribler use "fork-centric workflow" for developing new features (described below).

I suggest to use:
* branch-centric workflow for team members
* fork-centric workflow for outside collaborators

## Current behaviour

If you want to develop a new feature, you should:
1. Fork a repo (or update a forked repo)
2. Create a branch in forked repo
3. Make a few commits
4. Push the branch
5. Make a PR to upstream repo from your fork.

## Desirable behaviour

Team members should create a feature branch in **main Tribler repo**. 
Workflow details will be described after discussion https://github.com/Tribler/tribler/issues/5569

## Branch vs Fork

Quotes from the [article](https://www.pluralsight.com/blog/software-development/the-definitive-guide-to-forks-and-branches-in-git) that describe the situation better than I could:

Forks

> Forks are best used: when the intent of the ‘split’ is to create a logically independent project, which may never reunite with its parent.

Branches

> Branches are best used: when they are created as temporary places to work through a feature, with the intent to merge the branch with the origin.

Disadvantages of a fork-centric workflow

> With fork-centric workflows, developers each have their own completely independent repository. This makes it difficult to see what everyone is working on unless you can see everyones fork in one place.

> If you’re using forks, these changes would live in different repos. This means that there is not true “collaboration space” for the team, just a canonical repo that everyone submits changes to when they are ready to make changes in master.

> With a branch-centric flow, all commits exist on one repository. If everyone is pushing to their branch, you have access to all changes happening within your code base.

> Increased Operational Risk: With a branch-centric workflow, developers can push their changes to a common repo frequently (say, at the end of every day). When other developers update their working copies, they will receive these branches.
 
> A unique danger of a fork-centric workflow, is if a developer works in isolation on their own repo. Although not always the case, it presents a slightly increased risk profile for certain situations, like if the developer leaves the company or is temporarily unavailable. The risk here is knowledge management (i.e. other developers won’t necessarily know where the work is happening).

## Articles

* https://www.pluralsight.com/blog/software-development/the-definitive-guide-to-forks-and-branches-in-git
* https://github.community/t/branch-vs-fork/1206
* https://support.atlassian.com/bitbucket-cloud/docs/branch-or-fork-your-repository

**Comment by qstokkink** [comment]
I abjectly reject branch-based workflows on the main repository.. in IPv8. So, I would assume everyone expects me to now go on a 10 page rant: prepare to be disappointed 😃 I don't oppose this in Tribler and here's why (using quotes from the articles from the O.P.).

**Branching on the main repository**
A branch-based way of working is a centralized workflow. This is fit for ["a small group of programmers who trust each other and are in close communication"](https://support.atlassian.com/bitbucket-cloud/docs/branch-or-fork-your-repository/), allowing you to have a ["a rapid iteration cycle"](https://support.atlassian.com/bitbucket-cloud/docs/branch-or-fork-your-repository/). ["This tends to be a better fit for private repositories"](https://www.pluralsight.com/blog/software-development/the-definitive-guide-to-forks-and-branches-in-git) and is something we can experiment with now that we have a team of dedicated full-time developers, to increase productivity. ["There’s no right or wrong way to do it if it achieves your goals"](https://github.community/t/branch-vs-fork/1206) and our goal has shifted towards quicker releases. We should just make sure to evaluate the efficacy of this approach at some point (if it's better or worse than our current flow).

**Why not in IPv8**
IPv8 is a networking library, not an application, that is still wire-format compatible with code from 6 years ago and does not need or want rapid iterations. Small changes may lead to epic bugs (e.g. #4171) and we ["want fine-grain control over merging"](https://support.atlassian.com/bitbucket-cloud/docs/branch-or-fork-your-repository/) to make sure our core protocols do not break (e.g. fuelling the "passport prototype"). As such, IPv8 uses the idea of blessed repositories (see https://git-scm.com/book/en/v2/Distributed-Git-Distributed-Workflows). Integration is only done by a dictator/manager (manifesting as code reviews on GitHub): no developer is trusted a priori. Developers can still work on the same feature by opening a PR and "allow edits from maintainers", but, most importantly, without ever touching the blessed repository. IPv8 is also cloned a lot and anything inflating the `.git` folder is therefore disallowed (exercise for the reader: look at the `.git` folder size of Tribler -- yes, that's permanent).

**tl;dr**
Different projects, different ways of working. Not for IPv8, but I don't see why wouldn't try this out for Tribler and see if it works.

**Comment by ichorid** [comment]
We *love* to squash commits and rewrite history in local branches. Wouldn't that interfere with other developers' work if we do it in the same repo? Also, I personally have a lot of experimental branches which I often don't merge in the main repo. We'll have to use a specific branch naming scheme based on developers' names to manage the sheer number of branches that is going to appear in the main repor. E.g. `ichorid/feature/foobar`

**Comment by qstokkink** [comment]
We would definitely have to stop squashing while developing. But, even for that, revert commits and fixup commits exist. You can hold off sanitizing/rewriting the branch's history until just before the final merge.

The branch explosion also worries me. But, I guess we'll just have to see if it's acceptable 🤷‍♂️

**Comment by drew2a** [comment]
Squash or not to squash... It was my main theme for the next improvement request 😆

**Comment by drew2a** [comment]
@ichorid to use a specific branch naming scheme is a good idea. 
I suggest just change it a bit, to get more clear ordering:

* `feature/drew2a/whatever`
* `feature/ichorid/fast_vpn`
* `feature/ichorid/foobar`
* `release/7.6`

**Comment by drew2a** [comment]
@xoriole @devos50 what do you think, guys?

**Comment by devos50** [comment]
I think both options are viable. I have read the advantages and disadvantages of both approaches and if I have to make a choice, I prefer the fork-based for the following reason. I believe that having all pending work on a single repository would pollute the list of branches, unnecessary inflating the amount of stuff I get on every `git fetch`. Personally, I have many branches for experimental changes/debugging and putting them all in the Tribler upstream repository would result in quite a list of branches. I think that the branches in the main Tribler repo should exclusively be for 1) releases, and 2) features where multiple developers are simultaneously working on.

A small comment on one of the disadvantages:

> This makes it difficult to see what everyone is working on unless you can see everyones fork in one place.

I'm not sure if we need a fine-grained overview of what everyone is working on. Right now, we have a basic idea of what everyone is doing/is focussing on. In the situation that a dev needs help with their code, I simply add their repository to my workspace with `git remote add`.

**Comment by egbertbouman** [comment]
Another vote for the fork-centric workflow, since I think that most of the features are likely to be programmed by a single programmer anyway, and moving to a branch-centric workflow would just clutter the main repo. If there is a feature that multiple people work on, they can always choose to create a branch on the main repo.

**Comment by xoriole** [comment]
My personal choice would also be fork-centric workflow. I like the freedom to work and experiment on my own fork (rewrite/rebase/squash/delete) and only publish final code as PR to the main repo.

**Comment by drew2a** [comment]
Okay, thank you all.

Let's not change anything.
